### PR TITLE
Feat/m2 4234 cl backbutton behaviour

### PR DESCRIPTION
**Pull request by BamMironov**
Implementation of the new logic described here:  https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/282722434/Activity+with+Conditional+Logic+-+Back+Button+Press+Logic+-+Web
in the scope of https://mindlogger.atlassian.net/jira/software/c/projects/M2/issues/M2-4234?filter=myopenissues

### CODE REVIEW

File: src/features/pass-survey/model/hooks/useActivityState.ts [hunk 1]
```diff
@@ -109,7 +111,7 @@ function useActivityState({
     });
   }
 
-  function removeAnswer(step: number) {
+  function resetAnswer(step: number) {
```
Comment (anq83) [inline src/features/pass-survey/model/hooks/useActivityState.ts:114]: let's rename to undoAnswer?

File: src/entities/conditional-logic/model/conditions.test.ts [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+  });
+});
+
+describe('[isEqualToValue] function', () => {
+  const emptyObject = {};
+  const sameValueTestCases = [
+    { a: 1, b: 1 },
+    { a: 'Hello', b: 'Hello' },
+    { a: 0, b: 0 },
+    { a: emptyObject, b: emptyObject },
```
Comment (anq83) [inline src/entities/conditional-logic/model/conditions.test.ts:117]: Are you testing equality by reference here? If so, then do not emphasize that object is empty here.

**Review by anq83 — CHANGES_REQUESTED** [review]

File: src/features/pass-survey/model/AnswerValidator.ts [hunk 3]
```diff
@@ -25,87 +26,56 @@ function AnswerValidator(params?: AnswerValidatorArgs) {
...
-      if (currentAnswer?.answer == null) {
-        return false;
-      }
-
-      const answer = currentAnswer.answer as Item[];
-
-      return answer.find(
-        answerItem => String(answerItem.value) === optionValue,
+      const answer = (currentAnswer?.answer as Maybe<Item[]>)?.map(item =>
+        item.value.toString(),
```
[the remark this answers is not in the document]
Comment (BamMironov) [reply src/features/pass-survey/model/AnswerValidator.ts:67]: Yes, correct. The system can not have a list of selected items filled with undefined or objects with value as undefined

File: src/features/pass-survey/model/AnswerValidator.ts [hunk 4]
```diff
@@ -25,87 +26,56 @@ function AnswerValidator(params?: AnswerValidatorArgs) {
...
-      if (currentAnswer?.answer == null) {
-        return false;
-      }
-
-      const answer = currentAnswer.answer as Item[];
-
-      return !answer.find(
-        answerItem => String(answerItem.value) === optionValue,
+      const answer = (currentAnswer?.answer as Maybe<Item[]>)?.map(item =>
+        item.value.toString(),
```
[the remark this answers is not in the document]
Comment (BamMironov) [reply src/features/pass-survey/model/AnswerValidator.ts:75]: Sure. The same

File: src/features/pass-survey/model/StepperUtils.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import { ActivityState } from '../lib';
+
+class StepperUtils {
+  private activityState: ActivityState;
+
+  constructor(activityState: ActivityState) {
+    this.activityState = activityState;
+  }
+
+  public getNextStepShift(direction: 'forwards' | 'backwards'): number {
```
[the remark this answers is not in the document]
Comment (BamMironov) [reply src/features/pass-survey/model/StepperUtils.ts:11]: No problem. Any naming suggestion?

File: src/features/pass-survey/model/StepperUtils.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+    let shift = 1;
+
+    const visibilityChecker = PipelineVisibilityChecker(items, answers);
+
+    while (true) {
+      let nextStep = direction === 'forwards' ? step + shift : step - shift;
+
+      if (nextStep > items.length - 1) {
+        shift = items.length;
```
[the remark this answers is not in the document]
Comment (BamMironov) [reply src/features/pass-survey/model/StepperUtils.ts:22]: This code has been there for a long time. I guess, this is a check for a unreal edge case when one somehow goes beyond the pipeline. Maybe, the correct way would be to assign "items.length - 1" instead

File: src/features/pass-survey/model/StepperUtils.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        break;
+      }
+
+      if (nextStep < 0) {
+        break;
```
[the remark this answers is not in the document]
Comment (BamMironov) [reply src/features/pass-survey/model/StepperUtils.ts:27]: Right. We can cover it with tests. Will do

File: src/features/pass-survey/model/hooks/useActivityState.ts (hunk 1 above)
Comment (BamMironov) [reply src/features/pass-survey/model/hooks/useActivityState.ts:114]: Sure

File: src/entities/conditional-logic/model/conditions.test.ts (hunk 2 above)
Comment (BamMironov) [reply src/entities/conditional-logic/model/conditions.test.ts:117]: Yes, you're right. Will change it

**Review by anq83 — APPROVED** [review]

**Pull request merged**